File: challenges/1999/1559.DetectCyclesIn2dGrid.go
package challenges

var dirs = []int{-1, 0, 1, 0, -1}

func containsCycle(grid [][]byte) bool {
	// use stack and depth
	h, w := len(grid), len(grid[0])
	if h*w < 4 {
		return false
	}

	visited := make([][]bool, h)
	depth := make([][]int, h)

	for i := range visited {
		visited[i] = make([]bool, w)
		depth[i] = make([]int, w)
	}

	stack := [][]int{}
	found := false

	for i := 0; i < h; i++ {
		for j := 0; j < w; j++ {
			if visited[i][j] {
				continue
			}

			found, stack = walk(grid, stack, depth, visited, i, j, h, w)

			if found {
				return true
			}

			depth[i][j] = 0
		}
	}

	return false
}

func walk(grid [][]byte, stack, depth [][]int, visited [][]bool, i, j, h, w int) (bool, [][]int) {
	depth[i][j] = 1
	visited[i][j] = true
	stack = append(stack, []int{i, j, 0, 1})
	target := grid[i][j]

	// fmt.Println("====> ", string(target), i, j)

	for len(stack) > 0 {
		curr := stack[len(stack)-1]
		x, y, dir, d := curr[0], curr[1], curr[2], curr[3]
		done := true

		visited[x][y] = true
		depth[x][y] = d

		// fmt.Println("at node:", x, y, dir, d)

		for k := dir; k < 4; k++ {
			x0, y0 := x+dirs[k], y+dirs[k+1]

			// out of the grid
			if x0 < 0 || y0 < 0 || x0 >= h || y0 >= w || grid[x0][y0] != target {
				continue
			}

			// we loop back onto a visited node
			if visited[x0][y0] {
				// if the node is in the path and has been visited, found and done
				if depth[x0][y0] != 0 && d-depth[x0][y0] >= 3 {

					// fmt.Println(x0, y0)
					// fmt.Println(visited)
					// fmt.Println(depth)

					return true, nil
				}

				// not the desired node or visited, continue
				continue
			}

			//todo: push this node to the stack, and continue
			stack = append(stack, []int{x0, y0, 0, d + 1})
			done = false
			curr[2] = k + 1

			break
		}

		// pop, and reset the depth if all other nodes from this node have been visited
		if done {
			depth[x][y] = 0
			stack = stack[:len(stack)-1]
		}
	}

	return false, stack
}

File: challenges/1999/1559-detect-cycles-in-2d-grid.py
'''
1559-detect-cycles-in-2d-grid
'''

from typing import List


class Solution:
  def containsCycle(self, grid: List[List[str]]) -> bool:
    seen = set()
    chain = {}
    m, n = len(grid), len(grid[0])

    def nxt(x0: int, y0: int, s: int):
      for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        x1, y1 = x0+dx, y0+dy
        if x1 < 0 or x1 >= m or y1 < 0 or y1 >= n:
          continue

        if grid[x1][y1] != grid[x0][y0]:
          continue

        if (x1, y1) in chain and s-chain[x1, y1] > 1:
          # print('found:', chain, (x0, y0), (x1, y1))
          return (x1, y1), True

        if (x1, y1) in seen:
          continue

        return (x1, y1), False

      return None, False

    def check(x: int, y: int) -> bool:
      curr = [(x, y, 0)]
      chain.clear()
      chain[x, y] = 0
      seen.add((x, y))

      while curr:
        x, y, s = curr[-1]
        nxt_move, done = nxt(x, y, s)
        if done:
          return True

        if nxt_move is None:
          curr.pop()
          del chain[x, y]
          continue

        seen.add(nxt_move)
        chain[nxt_move] = s+1
        curr.append((nxt_move[0], nxt_move[1], s+1))

      return False

    for x in range(m):
      for y in range(n):
        if check(x, y):
          return True

    return False
